core-bundle/src/EventListener/TwoFactorFrontendListener.php

```diff
@@ -0,0 +1,127 @@
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\Exception\UnauthorizedHttpException;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+
+class TwoFactorFrontendListener
+{
+    /** @var ContaoFramework */
+    private $framework;
+
+    /**
+     * @var ScopeMatcher
+     */
+    private $scopeMatcher;
+
+    /**
+     * @var TokenStorageInterface
+     */
+    private $tokenStorage;
+
+    public function __construct(ContaoFramework $framework, ScopeMatcher $scopeMatcher, TokenStorageInterface $tokenStorage)
+    {
+        $this->framework = $framework;
+        $this->scopeMatcher = $scopeMatcher;
+        $this->tokenStorage = $tokenStorage;
+    }
+
+    public function onKernelRequest(GetResponseEvent $event): void
+    {
+        $request = $event->getRequest();
+        $token = $this->tokenStorage->getToken();
+
+        // Check if is frontend request
+        if (!$this->scopeMatcher->isFrontendRequest($request)) {
+            return;
+        }
+
+        // Check if is a supported token
+        if (!$token instanceof TwoFactorToken && !$token instanceof UsernamePasswordToken) {
+            return;
+        }
+
+        $page = $request->attributes->get('pageModel');
+
+        // Check if actual page is available
+        if (!$page instanceof PageModel) {
+            return;
+        }
+
+        $user = $token->getUser();
+
+        // Check if FrontendUser
+        if (!$user instanceof FrontendUser) {
+            return;
+        }
+
+        /** @var PageModel $adapter */
+        $adapter = $this->framework->getAdapter(PageModel::class);
+
+        // Check if user has two-factor disabled but is enforced
+        if (!$user->useTwoFactor && $page->enforceTwoFactor) {
+            // Search for two-factor page
+            $twoFactorPage = $adapter->findPublishedById($page->twoFactorJumpTo);
+
+            if (!$twoFactorPage instanceof PageModel) {
+                throw new PageNotFoundException('No two-factor authentication page found.');
+            }
+
+            // Two-factor page found, quit
+            if ($page->id === $twoFactorPage->id) {
+                return;
+            }
+
+            // Redirect to two-factor page
+            $event->setResponse(new RedirectResponse($twoFactorPage->getAbsoluteUrl()));
+
+            return;
+        }
+
+        // Check if user has TwoFactorToken
+        if (!$token instanceof TwoFactorToken) {
+            return;
+        }
+
+        // Search 401 error page
+        $unauthorizedPage = $adapter->find401ByPid($page->rootId);
+
+        if (!$unauthorizedPage instanceof PageModel) {
+            return;
+        }
+
+        // Check if 401 error page is available
+        $redirect = $adapter->findPublishedById($unauthorizedPage->jumpTo);
```

Comment:
I think the 401 does not necessary have a redirect page, setting one is optional. Also, this does not really do anything with the 401 page, so why are we checking for it? Just to not throw an exception?

Reply:
I don't get your point here.